src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala

```diff
@@ -0,0 +1,231 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2019 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.fulcrumgenomics.commons.io
+
+import java.nio.file.Path
+
+import com.fulcrumgenomics.commons.CommonsDef
+import com.fulcrumgenomics.commons.util.LazyLogging
+import com.fulcrumgenomics.commons.CommonsDef._
+//import com.fulcrumgenomics.commons.io.Python.executable
+
+import scala.collection.mutable.ListBuffer
+import scala.util.Try
+
+/** Provides methods to run an executable. For example, methods to run this
+  * executable using provided arguments, test if this executable is available, etc.
+  */
+trait CommandLineTool extends LazyLogging {
+
+  /** The available number of processors at runtime. */
+  lazy val availableProcessors: Int = Runtime.getRuntime.availableProcessors
+
+  /** The name of the executable. */
+  val executable: String
+
+  /** Arguments for the executable.
+    *
+    * For example, to run `Rscript --version`:
+    * {{{
+    * scala> import com.fulcrumgenomics.commons.io._
+    * scala> val testArgs = Seq("--version")
+    * scala> val executable = "Rscript"
+    * scala> Rscript.execCommand(executable +: testArgs: _*)
+    * res1: scala.util.Try[scala.collection.mutable.ListBuffer[String]] = Success(ListBuffer(R scripting front-end version 3.5.1 (2018-07-02)))
+    * }}}
+    * */
+  val testArgs: Seq[String]
+
+  /** Exception class that holds onto the exit/status code of the executable. */
+  case class ToolException(status: Int) extends RuntimeException {
+    override def getMessage: String = s"$executable failed with exit code $status."
+  }
+
+  /** Executes a command and returns the stdout and stderr.
+    *
+    * @param command the command to be executed
+    */
+  def execCommand(command: String*): Try[ListBuffer[String]] = {
```

Comment:
I've found keeping the streams separate more useful for shelling out and collecting output. I've since modified to:

```scala
case class ProcessOutput(
  stdout: ListBuffer[String] = ListBuffer[String](), // Is this a bad idiom?
  stderr: ListBuffer[String] = ListBuffer[String]()
)

/** Executes a command and returns the stdout and stderr.
  *
  * @param command the command to be executed
  */
def execCommand(command: String*): Try[ProcessOutput] = {
  logger.debug("Executing: " + command.mkString(" "))
  Try {
    val output   = ProcessOutput()
    val process  = new ProcessBuilder(command: _*).redirectErrorStream(false).start()
    val outSink  = new AsyncStreamSink(process.getInputStream, s => output.stdout.append(s))
    val errSink  = new AsyncStreamSink(process.getErrorStream, s => output.stderr.append(s))
    val exitCode = process.waitFor()
    if (exitCode != 0) throw ToolException(exitCode)
    Seq(outSink, errSink).foreach(_.safelyClose())
    output
  }
}
```